lib/Sema/CSSolver.cpp

```diff
@@ -1722,28 +1721,43 @@ void ConstraintSystem::partitionDisjunction(
 
   // Now collect the overload choices that are defined within the type
   // that was designated in the operator declaration.
-  auto *designatedNominal = getOperatorDesignatedNominalType(Choices[0]);
-  if (designatedNominal) {
-    forEachChoice(Choices, [&](unsigned index, Constraint *constraint) -> bool {
-        auto *decl = constraint->getOverloadChoice().getDecl();
-        auto *funcDecl = cast<FuncDecl>(decl);
-
-        auto *parentDecl = funcDecl->getParent()->getAsDecl();
-        if (parentDecl == designatedNominal) {
-          definedInDesignatedType.push_back(index);
-          return true;
-        }
+  auto designatedNominalTypes = getOperatorDesignatedNominalTypes(Choices[0]);
+  if (!designatedNominalTypes.empty()) {
+    forEachChoice(
+        Choices, [&](unsigned constraintIndex, Constraint *constraint) -> bool {
+          auto *decl = constraint->getOverloadChoice().getDecl();
+          auto *funcDecl = cast<FuncDecl>(decl);
+
+          auto *parentDecl = funcDecl->getParent()->getAsDecl();
+          for (auto designatedTypeIndex : indices(designatedNominalTypes)) {
+            auto *designatedNominal =
+                designatedNominalTypes[designatedTypeIndex];
+            if (parentDecl == designatedNominal) {
+              if (designatedTypeIndex >= definedInDesignatedType.size())
+                definedInDesignatedType.resize(designatedTypeIndex + 1);
+              auto &constraints = definedInDesignatedType[designatedTypeIndex];
+              constraints.push_back(constraintIndex);
+              return true;
+            }
 
-        if (auto *extensionDecl = dyn_cast<ExtensionDecl>(parentDecl)) {
-          parentDecl = extensionDecl->getExtendedNominal();
-          if (parentDecl == designatedNominal) {
-            definedInExtensionOfDesignatedType.push_back(index);
-            return true;
+            if (auto *extensionDecl = dyn_cast<ExtensionDecl>(parentDecl)) {
+              parentDecl = extensionDecl->getExtendedNominal();
+              if (parentDecl == designatedNominal) {
```

Comment:
You can use a extract resize + add into a separate function to deduplicate logic here for types and extensions

Reply:
Yes, I'll do that in a follow-up.